src/charm.py

```diff
@@ -1182,14 +1201,30 @@ def _get_juju_secret_field(self, label: str, field: str) -> Optional[str]:
             The value of the field is returned, or `None` if the field does not
             exist.
 
-            If the secret does not exist, `None` is returned.
+            If the secret does not exist, or there is an error retrieving the secret, `None` is returned.
         """
-        try:
-            juju_secret = self.model.get_secret(label=label)
-        except SecretNotFoundError:
-            return None
-        content = juju_secret.get_content(refresh=True)
-        return content.get(field)
+        content = self._get_juju_secret_content(label)
+        return content.get(field) if content else None
+
+    def _get_juju_secret_fields(self, label: str, *fields: str) -> Tuple[str | None, ...]:
```

Comment:
That's something I tried to find a better approach for, and admittedly tried a few (way too many) different approaches with typing but wasn't able to find anything great. I spent way more time on this than it would seem :laughing: .

Unfortunately, if we get rid of `_get_juju_secret_field`, then the return type for the single field case would be a tuple, and you will get type errors unless you unpack the (single) element.

We could use just `_get_juju_secret_field`, but then we're retrieving the secret every time for each field and causing additional overhead.

This approach isn't perfect because a return type where the tuple length is based on the argument's tuple length isn't currently possible in the Python typing system. However, with this approach the type checker seems to just trust you, so the code still looks fairly clean when retrieving multiple fields, at the cost of some type checking validation. In other words: if you ask for 2 fields, but unpack 3 (or vice versa), the type checker won't warn you. In the first case, you'll get a runtime error. In the second case (vice versa), your second value will contain a tuple with two elements.

The alternative is to create some stubs which cover the most common cases (we can create a stub for 2, 3, 4 fields) which will fix the type checker, but where do you stop? This seemed the best bang-for-buck.

Reply:
Couldn't we have `existing_token, _ = self._get_juju_secret_fields(AUTOUNSEAL_TOKEN_SECRET_LABEL, "token")` in the `_get_autounseal_vault_token` function and completely get rid of `_get_juju_secret_field`?

Reply:
I actually thought that would work, and that's what I meant when I said "unless you unpack the element" -- but I just tried it, and it doesn't actually work:

```
  File "/home/darndt/git/canonical/tls/vault-k8s-operator/src/charm.py", line 1166, in _get_autounseal_vault_token
    existing_token, _ = self._get_juju_secret_fields(AUTOUNSEAL_TOKEN_SECRET_LABEL, "token")
    ^^^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)
```

Reply:
It feels like there must be a way to figure this out :laughing: I'll go ahead and approve.

Reply:
I know.. it bothers me too.. I already tried for a while to figure this out.

But, I might have an idea. I mentioned stubbing out above. I wonder what happens if I only stub out the single field case. I'll give that a shot.

Reply:
OK, I had way too much fun trying this. I'm still curious if there's a better way to do this, but this is one alternative: https://github.com/canonical/vault-k8s-operator/pull/417/commits/49f1df9aa218bdf718ac16acd5345ddabb747c76

In this change, we have `@overload` stubs for a single field, and two fields. They are inferred correctly by the type checker. Anything more than that, and it will just trust you (as before).

Reply:
@gruyaume Let me know if you like this more or less

Reply:
Humm I preferred the older version with the two simpler methods. While there was duplication, it was easier to read/understand.

Reply:
@gruyaume I made a couple iterations on this, and the first (the one I linked above) was definitely the hardest to follow (I used some really ugly `, /` notation). Can you confirm if you saw this version? 

https://github.com/canonical/vault-k8s-operator/blob/69674a9b725952dca5cb42a92a577815d8408627/src/charm.py#L1193-L1226

If you still prefer the old one, I'll revert to that and merge.

Reply:
Yes I did see the version you copied here. I feel like readability wise the first version with `_get_juju_secret_field` and `_get_juju_secret_fields` was better.

Reply:
OK, I reverted back to the version that was already approved, so I'll enable auto-merge as a re-review shouldn't be necessary.